### DISCUSSION TITLE
Passing Headers to the next/router fetch call to getServerSideProps

### DISCUSSION THREAD

**Discussion by jamiechong** [vercel/next.js#19453] [Help]
Scenario:

I run my Next app behind a reverse-proxy gateway (Fastly). I've configured the gateway to send a request header to my app, which I use in `getServerSideProps` as a type of _context_. I do this because I don't want this dynamic value to be exposed in the URL.

Everything works great on the initial SSR. However, when I click a next/router `Link` to navigate within my app, I want the same header to be forwarded with the XHR (fetch) request to `getServerSideProps` for the CSR. This is the specific `fetch` call that is made (inside `fetchRetry`. https://github.com/vercel/next.js/blob/master/packages/next/next-server/lib/router/router.ts#L448  

How can I make this happen elegantly? Or am I completely missing something fundamental? 

I've had to implement this ugly HOC work around so I can hijack the fetch call used for the CSR:

```
let _originalFetch = null;
export const withMarketHeaderForSSRFetch = PageComponent => props => {
  const { market } = props;
  const header = "x-market-context";

  if (market && typeof window !== "undefined" && !_originalFetch) {

    _originalFetch = window.fetch;

    window.fetch = function(url, options) {
      if (url.includes("/_next/data")) {    
        if (options?.headers) {
          options.headers[header] = market;
        } else {
          options = { ...options, headers: { [header]: market } }
        }
      }
      return _originalFetch.call(window, url, options);
    }
  }
  return <PageComponent {...props} />
}
```

**Comment by jamiechong** [comment]
Any thoughts on this from within the community? When I navigate between pages I basically want to be able to send custom headers to my getServerSideProps function.

**Comment by akellbl4** [comment]
Hi,
Your solution is quite good for the problem. I don't see any other way to do so.
You can do different implementations, but the target will be the same – replacing global fetch with yours.

**Reply by jamiechong** [reply]
Preferably an interface is exposed via next.config.js (for the router) or something that allows me to send custom headers along with the internal `fetch` call made by Next.

**Reply by bquay** [reply]
Also running into a situation where this would be super useful. Going to give this a try!
